Secure Invoice

raphael-adam opened the pull request:
Zahlungsart "Secure Invoice".

TODO:
- [x] Changelog und Version

https://plentymarkets.kanbanize.com/ctrl_board/75/cards/83326/details/

@plentymarkets/team-order-payment

danielmarx committed:
update lib

raphael-adam committed:
test fixed vat rate

danielmarx committed:
fix price

raphael-adam committed:
fix amounts

raphael-adam committed:
check if delivery address is the same as invoice address

raphael-adam committed:
set settleaccount to auto in capture

danielmarx committed:
extended payone-api for getinvoice call

raphael-adam committed:
add settings for own portal for secure invoice

danielmarx committed:
add get invoice for test

danielmarx committed:
add get invoice for test

danielmarx committed:
add get invoice for test

danielmarx committed:
cleanup

danielmarx committed:
cleanup

danielmarx committed:
cleanup

danielmarx committed:
cleanup

raphael-adam committed:
validate if settings are set

raphael-adam committed:
check if currency is euro

raphael-adam committed:
change text

raphael-adam committed:
change text

raphael-adam committed:
change text

raphael-adam committed:
set businessrelation field

MKanitz commented on resources/lib/PayoneApi/Request/GetInvoice/GetInvoice.php:
```diff
@@ -0,0 +1,97 @@
+    public function getAmount()
+    {
+        return null;
+    }
```
Warum `Getter` die immer `NULL` zurückliefern?

MKanitz commented on resources/lib/PayoneApi/Request/GetInvoice/RequestFactory.php:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * @param string $paymentMethod
+     * @param array $data
+     * @param null $referenceId
```
`$referenceId` sollte `int` oder `null` sein können.

MKanitz commented on resources/lib/PayoneApi/Request/GetInvoice/RequestFactory.php:
```diff
@@ -0,0 +1,46 @@
+    public static function create($paymentMethod, $data, $referenceId = null)
+    {
+        if ($paymentMethod != PaymentTypes::PAYONE_INVOICE_SECURE) {
+            throw new \Exception('get invoice only by secure invoice');
```
Eher 'Get invoice only for secure invoice'

MKanitz commented on src/Controllers/StatusController.php:
```diff
@@ -71,14 +83,24 @@ public function index()
         $this->logger->debug('Controller.Status', $this->request->all());
 
-        if ($this->request->get('key') != md5($this->config->get('key'))) {
+        if ($this->request->get('key') != md5($this->config->get(PayoneInvoiceSecurePaymentMethod::PAYMENT_CODE . '.key')) &&
+            $this->request->get('key') != md5($this->config->get('key'))) {
```
Warum die doppelte Prüfung, wenn der Wert für `key` und für `PayoneInvoiceSecurePaymentMethod::PAYMENT_CODE . '.key'` in der Config identisch ist?

danielmarx replied:
Die sind ja nicht identisch, secure invoice hat einen anderen key

MKanitz commented on src/Methods/PaymentValidator.php:
```diff
@@ -58,8 +59,9 @@ public function validate(PaymentAbstract $payment)
         }
 
         $billingAddress = $this->addressHelper->getBasketBillingAddress($this->basket);
-        $shippingAddress = $this->addressHelper->getBasketShippingAddress($this->basket);
+        $deliveryAddress = $this->addressHelper->getBasketShippingAddress($this->basket);
         if (!$billingAddress) {
+            // TODO: shouldn't this be 'return false'?
             return true;
         }
```
Ich denke auch, dass hier `false` zurückgeliefert werden sollte.

raphael-adam replied:
Ich hatte auch schon mit Daniel drüber geredet. Es ist halt die Frage, ob es nicht doch aus irgendeinem Grund absichtlich auf `true` stand. Aber uns ist auch nichts eingefallen.

MKanitz replied:
Das Ding ist halt, dass wenn die Rechungsadresse nicht gesetzt ist, auch keine der anderen Prüfungen mehr erfolgen würde.

danielmarx replied:
Ich denke ab eine Zeitpunkt x wird immer eine Rechnungsadresse vorhanden sein. Und spätestens zu diesen Zeitpunkt fliegt die Zahlungsart raus

MKanitz commented on src/Methods/PaymentValidator.php:
```diff
@@ -69,6 +71,18 @@ public function validate(PaymentAbstract $payment)
             return false;
         }
+        
+        if (!$payment->canHandleDifferingDeliveryAddress() && $deliveryAddress && $billingAddress->id != $deliveryAddress->id) {
```
Hier könnte `$billingAddress` aktuell aber noch nicht gesetzt sein. Wie oben angemerkt.

raphael-adam replied:
Wieso? Es wird in dem Fall doch `return true` ausgegeben.

MKanitz replied:
Ja natürlich, hast recht. 👍

MKanitz commented on src/Models/PaymentConfig/ApiCredentials.php:
```diff
@@ -22,10 +22,18 @@ public function __construct(
     }
 
     /**
+     * @param null $paymentCode
```
Typ sollte hier auch `int` oder `null` sein.

MKanitz commented on src/Services/PaymentDocuments.php:
```diff
@@ -0,0 +1,214 @@
+    private $logger;
+
+    /**
+     * PaymentCreation constructor.
```
"PaymentDocuments constructor"

MKanitz commented on src/Services/PaymentDocuments.php:
```diff
@@ -0,0 +1,214 @@
+        $this->importInvoice($orderId, $invoiceId, $getInvoiceResult->getBase64(), $invoiceDate);
+    }
+
+    /**
+     * Imports one invoice.
+     *
+     * @param int $orderId
+     * @param array $invoice
+     * @throws InvalidDocumentTypeException
+     * @throws \Plenty\Exceptions\ValidationException
+     */
+    private function importInvoice(int $orderId, string $invoiceNumber, string $content, string $invoiceDate)
```
Parameter in der Doku stimmen nicht mit der Funktion überein.

estafilarakis commented on resources/lib/doGetDocument.php:
```diff
@@ -0,0 +1,48 @@
+        'Request was : ' . json_encode($serializer->serialize($request), JSON_PRETTY_PRINT) . PHP_EOL .
+        'Response was: ' . json_encode($serializer->serialize($response), JSON_PRETTY_PRINT)
+    );
+    return $errorResponse->jsonSerialize();
```
Wieso wird hier das Objekt `$errorResponse` als JSON serialisiert und das Objekt weiter oben in Zeile 34 nicht, obwohl beide vom selben Typ `ClientErrorResponse` sind?

danielmarx committed:
cleanup

danielmarx committed:
cleanup

MKanitz commented on src/Services/PaymentDocuments.php:
```diff
@@ -0,0 +1,210 @@
+        $this->logger = $logger;
+    }
+
+    public function uploadDocument($txid, $invoiceId, $invoiceDate, $invoiceTotal)
```
Hier fehlt die Doku.

raphael-adam replied:
Doku finde ich auch notwendig, es ist ja nicht selbsterklärend, was die Methode macht. Und ich finde der Name passt nicht, es wird doch gar nichts hochgeladen. Vllt `addInvoiceToOrder()`?

raphael-adam committed:
add changelog and increase version

raphael-adam commented on resources/lib/PayoneApi/Response/ClientErrorResponse.php:
```diff
@@ -8,16 +8,16 @@
 {
     /** @var string */
-    private $message;
+    private $errorMessage;
```
Warum das @danielmarx ? 🤔

danielmarx replied:
Weil der Error Log nicht funktioniert hat, der Name muss der gleiche Namen wie die get function haben

danielmarx committed:
add docu

raphael-adam reviewed (commented):
Der Teil von @danielmarx sieht für mich sonst gut aus

danielmarx commented on config.json:
```diff
@@ -168,6 +168,18 @@
+          "required": false,
+          "label": "Config.keyLabel",
+          "options": {}
+        },
```
Ich denke, das man für SecureInvoice ein eignes Portal und Key benötigt sollte man dann noch mal in die Doku aufnehmen.

raphael-adam replied:
Das wäre dann was für @jawendel ? Auch der Ablauf, dass die Rechnungen runtergeladen und an die Aufträge gespeichert werden, sollte ja eigentlich irgendwie dokumentiert werden.

jawendel replied:
Ja, das wäre etwas, das aufgenommen werden sollte. Ich habe mich mit dem Thema aber noch nicht auseinandergesetzt, deswegen kann ich noch nicht mehr dazu sagen.

raphael-adam replied:
Hab mal eine Karte dafür aufgenommen: https://plentymarkets.kanbanize.com/ctrl_board/75/cards/119127/details/

jawendel replied:
Sehr gut, danke.

raphael-adam commented on src/Services/PaymentDocuments.php:
```diff
@@ -0,0 +1,217 @@
+     * @param float $invoiceTotal
+     * @throws \Exception
+     */
+    public function uploadDocument($txid, $invoiceId, $invoiceDate, $invoiceTotal)
```
Nochmal der Kommentar, es hatte sich eingeklappt:

Doku finde ich auch notwendig, es ist ja nicht selbsterklärend, was die Methode macht. Und ich finde der Name passt nicht, es wird doch gar nichts hochgeladen. Vllt addInvoiceToOrder()?

danielmarx commented on src/Models/PaymentConfig/ApiCredentials.php:
```diff
@@ -22,10 +22,18 @@ public function __construct(
-    public function getKey()
+    public function getKey($paymentCode = null)
     {
+        if ($paymentCode !== null) {
```
Macht es hier vielleicht Sinn, zu prüfen ob es SecureInvoice ist, dann muss nicht jedesmal für alle payment arten geprüft werden ob es extra Zugangsdaten gibt. Die extra Config wird ja nur bei SecureInvoice benötigt, und bei dieser *muss* ja eine extra PortalId und Key hinterlegt sein.

raphael-adam replied:
Könnte man natürlich machen. Aber meinst du die extra Prüfung bringt wirklich was? So könnte man es später halt leichter auch für andere Zahlungsarten erweitern.

danielmarx replied:
War nur ein Idee, muss natürlich nicht gemacht werden.

danielmarx committed:
add docu

danielmarx committed:
Merge branch 'plentysystems_support' into InvoiceSecure

# Conflicts:
#	meta/documents/changelog_de.md
#	meta/documents/changelog_en.md

jawendel committed:
Documentation: updated user guides

- User Guides (De and En) updated
- added: Hinweis zum Gesicherten Rechnungskauf

danielmarx commented:
@jawendel danke für die Anpassung des User-Guilds
